pkg/agent/plugin/workloadattestor/k8s/sigstore/sigstore.go

```diff
@@ -0,0 +1,426 @@
+	Signature BundleSignature              `json:"signature"`
+}
+
+type BundleBody struct {
+	APIVersion string     `json:"apiVersion"`
+	Kind       string     `json:"kind"`
+	Spec       BundleSpec `json:"spec"`
+}
+
+type SelectorsFromSignatures struct {
+	Subject        string
+	Content        string
+	LogID          string
+	IntegratedTime string
+}
+
+func New(cache Cache, logger hclog.Logger) Sigstore {
+	return &sigstoreImpl{
+		verifyFunction:             cosign.VerifyImageSignatures,
+		fetchImageManifestFunction: remote.Get,
+		checkOptsFunction:          DefaultCheckOpts,
+
+		rekorURL: url.URL{
+			Scheme: rekor.DefaultSchemes[0],
+			Host:   rekor.DefaultHost,
+			Path:   rekor.DefaultBasePath,
+		},
+		logger:        logger,
+		sigstorecache: cache,
+	}
+}
+
+func DefaultCheckOpts(rekorURL url.URL) *cosign.CheckOpts {
+	co := &cosign.CheckOpts{}
+
+	// Set the rekor client
+	co.RekorClient = rekor.NewHTTPClientWithConfig(nil, rekor.DefaultTransportConfig().WithBasePath(rekorURL.Path).WithHost(rekorURL.Host))
+
+	co.RootCerts = fulcio.GetRoots()
+
+	return co
+}
+
+func (s *sigstoreImpl) SetLogger(logger hclog.Logger) {
+	s.logger = logger
+}
+
+// FetchImageSignatures retrieves signatures for specified image via cosign, using the specified rekor server.
+// Returns a list of verified signatures, and an error if any.
+func (s *sigstoreImpl) FetchImageSignatures(ctx context.Context, imageName string) ([]oci.Signature, error) {
+	ref, err := name.ParseReference(imageName)
+	if err != nil {
+		message := fmt.Errorf("error parsing image reference: %w", err)
+		return nil, message
+	}
+
+	if _, err := s.ValidateImage(ref); err != nil {
+		message := fmt.Errorf("could not validate image reference digest: %w", err)
+		return nil, message
+	}
+
+	co := s.checkOptsFunction(s.rekorURL)
+	sigs, ok, err := s.verifyFunction(ctx, ref, co)
+	if err != nil {
+		message := fmt.Errorf("error verifying signature: %w", err)
+		return nil, message
+	}
+	if !ok {
+		return nil, fmt.Errorf("bundle not verified for %q", imageName)
+	}
+
+	return sigs, nil
+}
+
+// ExtractSelectorsFromSignatures extracts selectors from a list of image signatures.
+// returns a list of selector strings.
+func (s *sigstoreImpl) ExtractSelectorsFromSignatures(signatures []oci.Signature, containerID string) []SelectorsFromSignatures {
+	// Payload can be empty if the attestor fails to retrieve it
+	if signatures == nil {
+		return nil
+	}
+	var selectors []SelectorsFromSignatures
+	for _, sig := range signatures {
+		// verify which subject
+		sigSelectors := s.SelectorValuesFromSignature(sig, containerID)
+		if sigSelectors != nil {
+			selectors = append(selectors, *sigSelectors)
+		}
+	}
+	return selectors
+}
+
+// SelectorValuesFromSignature extracts selectors from a signature.
+// returns a list of selectors.
+func (s *sigstoreImpl) SelectorValuesFromSignature(signature oci.Signature, containerID string) *SelectorsFromSignatures {
```

Comment:
I'm worried about all logs here, we create several logs and keep function to continue, why it is done this way? may we make function to fail instead?